rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,320 @@
+    bag_reader->open(temp_so, blank_converter_options);
+    ROSBAG2_CPP_LOG_INFO_STREAM("Opened");
+    // Get metadata from the bag file
+    auto temp_metadata = bag_reader->get_metadata();
+
+    if (temp_metadata.starting_time < metadata_.starting_time) {
+      metadata_.starting_time = temp_metadata.starting_time;
+    }
+    // ROSBAG2_CPP_LOG_INFO_STREAM(
+    //  "Current duration: " + std::to_string(metadata_.duration.count()));
+    // ROSBAG2_CPP_LOG_INFO_STREAM(
+    //   "Incoming duration: " + std::to_string(temp_metadata.duration.count()));
+    metadata_.duration += temp_metadata.duration;
+    ROSBAG2_CPP_LOG_INFO_STREAM("New duration: " + std::to_string(metadata_.duration.count()));
+    metadata_.message_count += temp_metadata.message_count;
+
+    // Add the topic metadata
+    for (const auto & topic : temp_metadata.topics_with_message_count) {
+      auto found_topic = std::find_if(
+        metadata_.topics_with_message_count.begin(),
+        metadata_.topics_with_message_count.end(),
+        [&topic](const rosbag2_storage::TopicInformation & agg_topic)
+        {return topic.topic_metadata.name == agg_topic.topic_metadata.name;});
+      if (found_topic == metadata_.topics_with_message_count.end()) {
+        // It's a new topic. Add it.
+        metadata_.topics_with_message_count.emplace_back(topic);
+      } else {
+        // Merge in the new information
+        found_topic->message_count += topic.message_count;
+        if (topic.topic_metadata.offered_qos_profiles != "") {
+          found_topic->topic_metadata.offered_qos_profiles =
+            topic.topic_metadata.offered_qos_profiles;
+        }
+        if (topic.topic_metadata.serialization_format != "") {
+          found_topic->topic_metadata.serialization_format =
+            topic.topic_metadata.serialization_format;
+        }
+        if (topic.topic_metadata.type != "") {
+          found_topic->topic_metadata.type = topic.topic_metadata.type;
+        }
+      }
+    }
+
+    ROSBAG2_CPP_LOG_INFO("Closing database");
+    bag_reader->reset();   // Close the reader's storage
+  }
+  bag_reader.reset();   // Delete the reader
+}
+
+void Reindexer::reindex(const rosbag2_storage::StorageOptions & storage_options)
+{
+  ROSBAG2_CPP_LOG_INFO("Beginning Reindex Operation.");
+
+  // Identify all database files
+  base_folder_ = storage_options.uri;
+  auto files = get_database_files(base_folder_);
+  std::cout << "Finished getting database files\n";
+  if (files.empty()) {
+    throw std::runtime_error("No database files found for reindexing. Abort");
+  }
+
+  // Create initial metadata
+  init_metadata(files, storage_options);
+  ROSBAG2_CPP_LOG_INFO_STREAM("Completed init_metadata");
+
+  // Collect all metadata from database files
+  aggregate_metadata(files, storage_options);
+  ROSBAG2_CPP_LOG_INFO_STREAM("Completed aggregate_metadata");
+
+  // Perform final touch-up
+  finalize_metadata();
```

Comment:
you're already aggregating duration and message counts in `aggregate_metadata` - i don't see why a separate vaguely-named function is needed to add up the sizes?